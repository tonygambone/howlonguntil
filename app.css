body {
    background-color: #223;
    font-family: sans-serif;
    margin: 0;
    padding: 8px;
    color: white;
}

#container {
    display: flex;
    flex-direction: row;
    flex-wrap: wrap;
    align-items: stretch;
    justify-content: space-between;
}

div.event {
    border: 1px solid #99a;
    box-sizing: border-box;
    flex-basis: auto;
    width: 49%;
    padding: 8px;
    margin: 8px 0;
}

div.event > h3 {
    margin-top: 0;
}