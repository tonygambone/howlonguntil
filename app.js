'use strict';

(function() {
    document.addEventListener("DOMContentLoaded", function(e) {
        // update duration displays
        window.setInterval(function() {
            var elements = document.querySelectorAll('.until[data-target]');
            elements.forEach(function(el) {
                var target = el.getAttribute('data-target');
                el.innerHTML = timeRemaining(target);
            });
        }, 1000);

        // initial setup
        refresh();
    });

    // pull the data and rebuild the UI
    function refresh() {
        fetch('events.json')
            .then(function(response) {
                return response.json();
            })
            .then(function(events) {
                // calculate next instance and sort events
                events = events.sort(function(a, b) {
                    a._nextInstance = a._nextInstance || getNextInstance(a);
                    b._nextInstance = b._nextInstance || getNextInstance(b);
                    return a._nextInstance.valueOf() - b._nextInstance.valueOf();
                });

                var container = document.getElementById('container');
                container.innerHTML = "";
                events.forEach(function (evt) {
                    // default display format
                    evt.displayFormat = evt.displayFormat || "dddd, MMMM Do, YYYY";

                    // hide past events after a certain time
                    evt.hideAfter = evt.hideAfter || [1, 'day'];
                    if (evt._nextInstance.diff(moment(), evt.hideAfter[1]) <= -evt.hideAfter[0]) return;

                    // build dom elements
                    var el = document.createElement('div');
                    el.setAttribute('class', 'event');

                    var title = document.createElement('h3');
                    title.innerHTML = evt.name;
                    el.appendChild(title);

                    var ts = document.createElement('div');
                    ts.setAttribute('class', 'timestamp');
                    ts.innerHTML = evt._nextInstance.format(evt.displayFormat);
                    el.appendChild(ts);

                    var until = document.createElement('div');
                    until.setAttribute('class', 'until');
                    until.setAttribute('data-target', evt._nextInstance.toISOString());
                    until.innerHTML = timeRemaining(evt._nextInstance);
                    el.appendChild(until);

                    container.appendChild(el);
                });
            });
    }

    // calculate the next instance of an event
    function getNextInstance(evt) {
        var startTime = moment(evt.startTime);
        switch (evt.repeat) {
            case "once":
                break;
            case "year":
            case "month":
            case "week":
            case "day":
            case "hour":
            case "minute":
            case "second":
                while (startTime.diff() < 0) {
                    startTime.add(1, evt.repeat);
                }
        }
        return startTime;
    }

    // get a string that shows the remaining time (or time since)
    function timeRemaining(target) {
        var diff = moment(target).diff(moment(), 'ms', true);
        if (diff <= 0 && diff >= -2000) {
            refresh();
        }
        var str = moment.duration(diff).format('d [days], h [hours], m [minutes], s [seconds]');
        str = str.replace(/(^| )(1 )(.+?)s/g, '$1$2$3'); // singular
        if (diff < 0) {
            str = str.replace(/^-?(.*)$/, '$1 ago');
        }
        return str;
    }
})();