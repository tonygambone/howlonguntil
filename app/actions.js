
const Actions = {
    FETCH: 'FETCH',
    EVENTS_RECEIVED: 'EVENTS_RECEIVED',
    TICK: 'TICK'
};

export const fetchEvents = function() {
    return {
        type: Actions.FETCH
    };
};

export default Actions;