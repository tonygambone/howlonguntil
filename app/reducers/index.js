
import Actions from '../actions';

const initialState = { events: [], elapsed: 0 };

export default (state = initialState, action) => {
    if (!state) { state = initialState };
    switch (action.type) {
        case Actions.EVENTS_RECEIVED:
            return {
                ...state,
                events: action.events,
                refreshTimeout: action.refreshTimeout
            };
        case Actions.TICK:
            return {
                ...state,
                elapsed: state.elapsed + 1
            };
        default:
            return state;
    }
};