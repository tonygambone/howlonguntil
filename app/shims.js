
// ensure the duration format is present whenever
// moment is imported
import moment from 'moment';
require('moment-duration-format');

export { moment };