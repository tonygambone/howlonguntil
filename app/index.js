
import React from 'react';
import { render } from 'react-dom';
import { Provider } from 'react-redux';
import { createStore, applyMiddleware } from 'redux';
import { App } from './components/app';
import reducer from './reducers';
import Actions from './actions';
import dataService from './services/data';
import eventService from './services/event';
import './app.css';

const store = createStore(
    reducer,
    undefined,
    applyMiddleware(dataService, eventService)
);

render(
    <Provider store={store}>
        <App title="How Long Until?" />
    </Provider>,
    document.getElementById('app')
);

store.dispatch({ type: Actions.FETCH });

setInterval(() => {
    store.dispatch({ type: Actions.TICK });
}, 1000);