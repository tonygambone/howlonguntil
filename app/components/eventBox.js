import React from 'react';
import { moment } from '../shims';

// get a string that shows the remaining time (or time since)
function timeRemaining(target) {
    var diff = moment(target).diff(moment(), 'ms', true);
    var str = moment.duration(diff).format('d [days], h [hours], m [minutes], s [seconds]');
    str = str.replace(/(^| )(1 )(.+?)s/g, '$1$2$3'); // singular
    if (diff < 0) {
        str = str.replace(/^-?(.*)$/, '$1 ago');
    }
    return str;
}

export class EventBox extends React.Component {
    render() {
        return (
            <div className="event">
                <h3>{this.props.event.name}</h3>
                <div className="timestamp">{this.props.event.nextInstance.format(this.props.event.displayFormat)}</div>
                <div className="until">{timeRemaining(this.props.event.nextInstance)}</div>
            </div>
        )
    }
}