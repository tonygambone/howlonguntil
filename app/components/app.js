import React from 'react';
import { connect } from 'react-redux';
import { EventBox } from './eventBox';

class AppInternal extends React.Component {
    render() {
        var eventBoxes = this.props.events.map((e) => <EventBox event={ e } />);
        return (
            <div>
                <h1>{this.props.title}</h1>
                <div id="container">
                    {eventBoxes}
                </div>
            </div>
        );
    }
}

export const App = connect(
    function mapStateToProps(state) {
        return { events: state.events };
    }
)(AppInternal);