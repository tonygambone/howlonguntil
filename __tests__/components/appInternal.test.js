import { AppInternal } from '../../app/components/app';
import renderer from 'react-test-renderer';
import React from 'react';
import { shallow } from 'enzyme';
import { fetchEvents } from '../../app/actions';

describe('the internal app component', () => {
    it('fetches events when the header is clicked', () => {
        let dispatch = jest.fn();
        const app = shallow(
            <AppInternal title="test title" events={[{id:1}, {id:2}, {id:3}]} elapsed="0" dispatch={dispatch} />
        );
        app.find('h1').simulate('click');
        expect(dispatch).toHaveBeenCalledTimes(1);
        expect(dispatch).toHaveBeenCalledWith(fetchEvents());
    });

    it('renders an eventbox for each event', () => {
        const app = shallow(
            <AppInternal title="test title" events={[{id:1}, {id:2}, {id:3}]} elapsed="0" />
        );
        let children = app.find('#container').children();
        expect(children.length).toEqual(3);
        for (let i = 0; i < 3; i++) {
            expect(children.at(i).key()).toBe((i+1).toString());
            expect(children.at(i).prop('event')).toBeDefined();
            expect(children.at(i).prop('elapsed')).toBeDefined();
        }
    });

    it('renders the header text', () => {
        const app = shallow(
            <AppInternal title="test title" events={[]} elapsed="0" />
        );
        expect(app.find('h1').text()).toEqual('test title');
    });

    it('matches the basic snapshot', () => {
        const app = renderer.create(
            <AppInternal title="test title" events={[]} elapsed="0" />
        );
        let tree = app.toJSON();
        expect(tree).toMatchSnapshot();
    });
});