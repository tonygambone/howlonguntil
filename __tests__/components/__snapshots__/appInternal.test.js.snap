exports[`the internal app component matches the basic snapshot 1`] = `
<div>
  <h1
    onClick={[Function]}>
    test title
  </h1>
  <div
    id="container" />
</div>
`;
